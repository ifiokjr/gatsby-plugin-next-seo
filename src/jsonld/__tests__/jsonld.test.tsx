import { assertSchema } from '@cypress/schema-tools';
import { render as testRender } from '@testing-library/react';
import mockConsole, { RestoreConsole } from 'jest-mock-console';
import React, { ReactElement } from 'react';
import { HelmetProvider } from 'react-helmet-async';

import {
  ArticleJsonLd,
  BlogJsonLd,
  BlogPostJsonLd,
  BookJsonLd,
  BreadcrumbJsonLd,
  CourseJsonLd,
  FAQJsonLd,
  LocalBusinessJsonLd,
  LogoJsonLd,
  NewsArticleJsonLd,
  ProductJsonLd,
  SpeakableJsonLd,
} from '../..';
import schemas from '../../../e2e/schema';

const render = (ui: ReactElement) =>
  testRender(ui, { wrapper: HelmetProvider });

let restoreConsole: RestoreConsole;

beforeAll(() => {
  restoreConsole = mockConsole('error');
});

afterAll(() => {
  restoreConsole();
});

test('ArticleJsonLd', () => {
  render(
    <ArticleJsonLd
      url='https://example.com/article'
      headline='Article headline'
      images={[
        'https://example.com/photos/1x1/photo.jpg',
        'https://example.com/photos/4x3/photo.jpg',
        'https://example.com/photos/16x9/photo.jpg',
      ]}
      datePublished='2015-02-05T08:00:00+08:00'
      dateModified='2015-02-05T09:00:00+08:00'
      authorName='Jane Blogs'
      publisherName='Ifiok Jr.'
      publisherLogo='https://www.example.com/photos/logo.jpg'
      description='This is a mighty good description of this article.'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Article', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('NewsArticleJsonLd', () => {
  render(
    <NewsArticleJsonLd
      url='https://example.com/newsarticle'
      title='News Article headline'
      images={[
        'https://example.com/photos/1x1/photo.jpg',
        'https://example.com/photos/4x3/photo.jpg',
        'https://example.com/photos/16x9/photo.jpg',
      ]}
      section='politics'
      keywords='prayuth, taksin, thai'
      dateCreated='2015-02-05T08:00:00+08:00'
      datePublished='2015-02-05T08:00:00+08:00'
      dateModified='2015-02-05T09:00:00+08:00'
      authorName='Jane Blogs'
      publisherName='Ifiok Jr.'
      publisherLogo='https://www.example.com/photos/logo.jpg'
      description='This is a mighty good description of this news article.'
      body='This is article body of news article'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('NewsArticle', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('BlogPostJsonLd', () => {
  render(
    <BlogPostJsonLd
      url='https://example.com/blog'
      headline='Blog headline'
      images={[
        'https://example.com/photos/1x1/photo.jpg',
        'https://example.com/photos/4x3/photo.jpg',
        'https://example.com/photos/16x9/photo.jpg',
      ]}
      datePublished='2015-02-05T08:00:00+08:00'
      dateModified='2015-02-05T09:00:00+08:00'
      authorName='Jane Blogs'
      description='This is a mighty good description of this blog.'
      publisherName='Ifiok Jr.'
      publisherLogo='https://www.example.com/photos/logo.jpg'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('BlogPosting', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('BlogJsonLd', () => {
  render(
    <BlogJsonLd
      url='https://example.com/blog'
      headline='Blog headline'
      images={[
        'https://example.com/photos/1x1/photo.jpg',
        'https://example.com/photos/4x3/photo.jpg',
        'https://example.com/photos/16x9/photo.jpg',
      ]}
      posts={[{ headline: 'Post 1' }, { headline: 'Post 2' }]}
      datePublished='2015-02-05T08:00:00+08:00'
      dateModified='2015-02-05T09:00:00+08:00'
      authorName='Jane Blogs'
      description='This is a mighty good description of this blog.'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Blog', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('BreadcrumbJsonLd', () => {
  render(
    <BreadcrumbJsonLd
      itemListElements={[
        {
          position: 1,
          name: 'Books',
          item: 'https://example.com/books',
        },
        {
          position: 2,
          name: 'Authors',
          item: 'https://example.com/books/authors',
        },
        {
          position: 3,
          name: 'Ann Leckie',
          item: 'https://example.com/books/authors/annleckie',
        },
        {
          position: 4,
          name: 'Ancillary Justice',
          item: 'https://example.com/books/authors/annleckie/ancillaryjustice',
        },
      ]}
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Breadcrumb', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('BookJsonLd', () => {
  render(
    <BookJsonLd
      author={{ name: 'Tolu B.' }}
      url='https://example.com/tolub'
      name='Rock your world - the final chapter'
      workExample={[
        {
          bookFormat: 'AudiobookFormat',
          isbn: '123123123',
          potentialAction: {
            expectsAcceptanceOf: {
              '@type': 'Offer',
              price: '6.99',
              priceCurrency: 'USD',
              eligibleRegion: {
                '@type': 'Country',
                name: 'US',
              },
              availability: 'http://schema.org/InStock',
            },
            target: {
              '@type': 'EntryPoint',
              urlTemplate:
                'http://www.barnesandnoble.com/store/info/offer/0316769487?purchase=true',
              actionPlatform: [
                'http://schema.org/DesktopWebPlatform',
                'http://schema.org/IOSPlatform',
                'http://schema.org/AndroidPlatform',
              ],
            },
          },
        },
      ]}
    />,
  );
  JSON.parse(document.querySelector('script')?.innerHTML!);
  expect(document.documentElement).toMatchSnapshot();
});

test('CourseJsonLd', () => {
  render(
    <CourseJsonLd
      name='Course Name'
      providerName='Course Provider'
      providerUrl='https//www.example.com/provider'
      description='Course description goes right here'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Course', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('SpeakableJsonLd', () => {
  render(<SpeakableJsonLd cssSelector={['#abc', '#root']} />);
  JSON.parse(document.querySelector('script')?.innerHTML!);
  expect(document.documentElement).toMatchSnapshot();
  expect(() =>
    render(
      <SpeakableJsonLd cssSelector={['#abc', '#root']} xpath={['/asdf/a']} />,
    ),
  ).toThrowError();
});

test('LocalBusinessJsonLd', () => {
  render(
    <LocalBusinessJsonLd
      type='Store'
      id='http://davesdeptstore.example.com'
      name="Dave's Department Store"
      description="Dave's latest department store in San Jose, now open"
      url='http://www.example.com/store-locator/sl/San-Jose-Westgate-Store/1427'
      telephone='+14088717984'
      address={{
        streetAddress: '1600 Saratoga Ave',
        addressLocality: 'San Jose',
        addressRegion: 'CA',
        postalCode: '95129',
        addressCountry: 'US',
      }}
      geo={{
        latitude: '37.293058',
        longitude: '-121.988331',
      }}
      images={[
        'https://example.com/photos/1x1/photo.jpg',
        'https://example.com/photos/4x3/photo.jpg',
        'https://example.com/photos/16x9/photo.jpg',
      ]}
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Local Business', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('LogoJsonLd', () => {
  render(
    <LogoJsonLd
      logo='http://www.your-site.com/images/logo.jpg'
      url='http://www.your-site.com'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Logo', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('ProductJsonLd', () => {
  render(
    <ProductJsonLd
      productName='Executive Anvil'
      images={[
        'https://example.com/photos/1x1/photo.jpg',
        'https://example.com/photos/4x3/photo.jpg',
        'https://example.com/photos/16x9/photo.jpg',
      ]}
      description="Sleeker than ACME's Classic Anvil, the Executive Anvil is perfect for the business traveler looking for something to drop from a height."
      brand='ACME'
      reviews={[
        {
          author: 'Jim',
          datePublished: '2017-01-06T03:37:40Z',
          reviewBody:
            'This is my favorite product yet! Thanks Nate for the example products and reviews.',
          name: 'So awesome!!!',
          reviewRating: {
            bestRating: '5',
            ratingValue: '5',
            worstRating: '1',
          },
        },
      ]}
      aggregateRating={{
        ratingValue: '4.4',
        reviewCount: 89,
      }}
      offers={{
        price: '119.99',
        priceCurrency: 'USD',
        priceValidUntil: '2020-11-05',
        itemCondition: 'UsedCondition',
        availability: 'InStock',
        url: 'https://www.example.com/executive-anvil',
        seller: {
          name: 'Executive Objects',
        },
      }}
      mpn='925872'
    />,
  );
  const jsonLD = JSON.parse(
    document.querySelector('script')?.innerHTML ?? '{}',
  );
  assertSchema(schemas)('Product', '1.0.0')(jsonLD);
  expect(document.documentElement).toMatchSnapshot();
});

test('FAQJsonLd', () => {
  render(
    <FAQJsonLd
      questions={[
        { question: 'What?', answer: 'Stand' },
        { question: 'How?', answer: 'Effort' },
        { question: 'Why?', answer: 'Peace' },
      ]}
    />,
  );
  JSON.parse(document.querySelector('script')?.innerHTML!);
  expect(document.documentElement).toMatchSnapshot();
});
