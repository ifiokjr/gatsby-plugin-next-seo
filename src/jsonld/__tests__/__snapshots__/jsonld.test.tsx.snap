// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Article",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://example.com/article"
  },
  "headline": "Article headline",
  "image": [
    "https://example.com/photos/1x1/photo.jpg",
    "https://example.com/photos/4x3/photo.jpg",
    "https://example.com/photos/16x9/photo.jpg"
  ],
  "datePublished": "2015-02-05T08:00:00+08:00",
  "dateModified": "2015-02-05T09:00:00+08:00",
  "author": {
    "@type": "Person",
    "name": "Jane Blogs"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Ifiok Jr.",
    "logo": {
      "@type": "ImageObject",
      "url": "https://www.example.com/photos/logo.jpg"
    }
  },
  "description": "This is a mighty good description of this article."
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`BlogJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Blog",
  "mainEntityOfPage": {
    "@type": "Blog",
    "@id": "https://example.com/blog"
  },
  "headline": "Blog headline",
  "image": [
    "https://example.com/photos/1x1/photo.jpg",
    "https://example.com/photos/4x3/photo.jpg",
    "https://example.com/photos/16x9/photo.jpg"
  ],
  "datePublished": "2015-02-05T08:00:00+08:00",
  "dateModified": "2015-02-05T09:00:00+08:00",
  "description": "This is a mighty good description of this blog.",
  "author": {
    "@type": "Person",
    "name": "Jane Blogs"
  },
  "blogPost": [
    {
      "@type": "BlogPosting",
      "headline": "Post 1"
    },
    {
      "@type": "BlogPosting",
      "headline": "Post 2"
    }
  ]
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`BlogPostJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://example.com/blog"
  },
  "headline": "Blog headline",
  "image": [
    "https://example.com/photos/1x1/photo.jpg",
    "https://example.com/photos/4x3/photo.jpg",
    "https://example.com/photos/16x9/photo.jpg"
  ],
  "datePublished": "2015-02-05T08:00:00+08:00",
  "dateModified": "2015-02-05T09:00:00+08:00",
  "author": {
    "@type": "Person",
    "name": "Jane Blogs"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Ifiok Jr.",
    "logo": {
      "@type": "ImageObject",
      "url": "https://www.example.com/photos/logo.jpg"
    }
  },
  "description": "This is a mighty good description of this blog."
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`BookJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Book",
  "name": "Rock your world - the final chapter",
  "url": "https://example.com/tolub",
  "author": {
    "@type": "Person",
    "name": "Tolu B."
  },
  "workExample": [
    {
      "@type": "Book",
      "bookFormat": "https://schema.org/AudiobookFormat",
      "isbn": "123123123",
      "person": {
        "@type": "Person"
      },
      "potentialAction": {
        "@type": "ReadAction",
        "expectsAcceptanceOf": {
          "@type": "Offer",
          "price": "6.99",
          "priceCurrency": "USD",
          "eligibleRegion": {
            "@type": "Country",
            "name": "US"
          },
          "availability": "http://schema.org/InStock"
        },
        "target": {
          "@type": "EntryPoint",
          "urlTemplate": "http://www.barnesandnoble.com/store/info/offer/0316769487?purchase=true",
          "actionPlatform": [
            "http://schema.org/DesktopWebPlatform",
            "http://schema.org/IOSPlatform",
            "http://schema.org/AndroidPlatform"
          ]
        }
      }
    }
  ]
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`BreadcrumbJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://example.com/books",
        "name": "Books",
        "@type": "Thing"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "https://example.com/books/authors",
        "name": "Authors",
        "@type": "Thing"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "https://example.com/books/authors/annleckie",
        "name": "Ann Leckie",
        "@type": "Thing"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "https://example.com/books/authors/annleckie/ancillaryjustice",
        "name": "Ancillary Justice",
        "@type": "Thing"
      }
    }
  ]
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`CourseJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Course",
  "name": "Course Name",
  "description": "Course description goes right here",
  "provider": {
    "@type": "Organization",
    "name": "Course Provider",
    "sameAs": "https//www.example.com/provider"
  }
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`FAQJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "FAQPage",
  "mainEntity": [
    {
      "@type": "Question",
      "acceptedAnswer": {
        "@type": "Answer",
        "text": "Stand"
      },
      "name": "What?"
    },
    {
      "@type": "Question",
      "acceptedAnswer": {
        "@type": "Answer",
        "text": "Effort"
      },
      "name": "How?"
    },
    {
      "@type": "Question",
      "acceptedAnswer": {
        "@type": "Answer",
        "text": "Peace"
      },
      "name": "Why?"
    }
  ]
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`LocalBusinessJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Store",
  "@id": "http://davesdeptstore.example.com",
  "name": "Dave's Department Store",
  "description": "Dave's latest department store in San Jose, now open",
  "url": "http://www.example.com/store-locator/sl/San-Jose-Westgate-Store/1427",
  "telephone": "+14088717984",
  "image": [
    "https://example.com/photos/1x1/photo.jpg",
    "https://example.com/photos/4x3/photo.jpg",
    "https://example.com/photos/16x9/photo.jpg"
  ],
  "geo": {
    "@type": "GeoCoordinates",
    "latitude": "37.293058",
    "longitude": "-121.988331"
  },
  "address": {
    "@type": "PostalAddress",
    "streetAddress": "1600 Saratoga Ave",
    "addressLocality": "San Jose",
    "addressRegion": "CA",
    "postalCode": "95129",
    "addressCountry": "US"
  }
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`LogoJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Organization",
  "url": "http://www.your-site.com",
  "logo": "http://www.your-site.com/images/logo.jpg"
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`NewsArticleJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://example.com/newsarticle"
  },
  "headline": "News Article headline",
  "image": [
    "https://example.com/photos/1x1/photo.jpg",
    "https://example.com/photos/4x3/photo.jpg",
    "https://example.com/photos/16x9/photo.jpg"
  ],
  "datePublished": "2015-02-05T08:00:00+08:00",
  "dateModified": "2015-02-05T09:00:00+08:00",
  "dateCreated": "2015-02-05T08:00:00+08:00",
  "author": {
    "@type": "Person",
    "name": "Jane Blogs"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Ifiok Jr.",
    "logo": {
      "@type": "ImageObject",
      "url": "https://www.example.com/photos/logo.jpg"
    }
  },
  "description": "This is a mighty good description of this news article.",
  "articleBody": "This is article body of news article",
  "keywords": "prayuth, taksin, thai",
  "articleSection": "politics"
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`ProductJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "Product",
  "name": "Executive Anvil",
  "image": [
    "https://example.com/photos/1x1/photo.jpg",
    "https://example.com/photos/4x3/photo.jpg",
    "https://example.com/photos/16x9/photo.jpg"
  ],
  "mpn": "925872",
  "brand": {
    "@type": "Brand",
    "name": "ACME"
  },
  "description": "Sleeker than ACME's Classic Anvil, the Executive Anvil is perfect for the business traveler looking for something to drop from a height.",
  "review": [
    {
      "@type": "Review",
      "author": "Jim",
      "datePublished": "2017-01-06T03:37:40Z",
      "reviewBody": "This is my favorite product yet! Thanks Nate for the example products and reviews.",
      "name": "So awesome!!!",
      "reviewRating": {
        "@type": "Rating",
        "bestRating": "5",
        "ratingValue": "5",
        "worstRating": "1"
      }
    }
  ],
  "aggregateRating": {
    "@type": "AggregateRating",
    "ratingValue": "4.4",
    "reviewCount": 89
  },
  "offers": {
    "@type": "Offer",
    "price": "119.99",
    "priceCurrency": "USD",
    "priceValidUntil": "2020-11-05",
    "itemCondition": "http://schema.org/UsedCondition",
    "availability": "http://schema.org/InStock",
    "url": "https://www.example.com/executive-anvil",
    "seller": {
      "@type": "Organization",
      "name": "Executive Objects"
    }
  }
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;

exports[`SpeakableJsonLd 1`] = `
<html>
  <head>
    <script
      data-rh="true"
      type="application/ld+json"
    >
      {
  "@context": "https://schema.org",
  "@type": "SpeakableSpecification",
  "cssSelector": [
    "#abc",
    "#root"
  ]
}
    </script>
  </head>
  <body>
    <div />
  </body>
</html>
`;
